awx_collection/tests/integration/targets/tower_schedule_rrule/tasks/main.yml

```diff
@@ -0,0 +1,252 @@
+
+- name: Test month_day_number > 31
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', month_day_number='32') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is failed
+      - '"month_day_number must be between 1 and 31" in result.msg'
+
+- name: Test valid month_day_number
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', month_day_number='18') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is not failed
+      - result.msg == 'DTSTART;TZID=America/New_York:20200416T034507 RRULE:FREQ=MONTHLY;BYMONTHDAY=18;INTERVAL=1'
+
+- name: Test on_the as junk
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', on_the='junk') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is failed
+      - '"on_the parameter must be two space seperated words" in result.msg'
+
+- name: Test on_the with invalid occurance
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', on_the='junk wednesday') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is failed
+      - '"The first string of the on_the parameter is not valid" in result.msg'
+
+- name: Test on_the with invalid weekday
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', on_the='second junk') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is failed
+      - '"Weekday portion of on_the parameter is not valid" in result.msg'
+
+- name: Test a valid on_the
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', on_the='second sunday') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is not failed
+      - result.msg == 'DTSTART;TZID=America/New_York:20200416T034507 RRULE:FREQ=MONTHLY;BYSETPOS=2;BYDAY=SU;INTERVAL=1'
+
+- name: Test an invalid timezone
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', timezone='junk') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is failed
+      - "'Timezone parameter is not valid' in result.msg"
+
+- name: Test an valid timezone
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month', start_date='2020-4-16 03:45:07', timezone='Zulu') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is not failed
+      - result.msg == 'DTSTART;TZID=Zulu:20200416T034507 RRULE:FREQ=MONTHLY;INTERVAL=1'
+
+- name: Test an empty schedule
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'month') }}"
+  ignore_errors: true
+  register: result
+
+- assert:
+    that:
+      - result is not failed
```

Comment:
these are all fine assertions to keep around for rrule testing, but these are just really high-level tests, and not a good place to put them.

As for the _best_ place for them, it seems like this would be ideal:

https://github.com/ansible/awx/blob/29b567d6e16b620b0eedce1ff8ee2c53828758dd/awx/main/tests/functional/api/test_schedules.py#L128

We are rapid-fire firing off mock requests and asserting that we get the right validation error. These integration tests for the collection do the same thing but in a very verbose way. And honestly these should be unit, not functional tests too, because the database should not be needed either.

So I would recommend that we add 1 test here that hits a 400 error, assert that the right error message is in `result.msg`. Then that should cover all of these cases, as far as the collection is concerned. The specific message is different, but the mechanics on the side of the collection are the same for all.

Reply:
oh wait, I wasn't fully understanding, this is a fully independent implementation of rrule logic...